wildcard/matcher.go

```diff
@@ -0,0 +1,159 @@
+
+var (
+	errEmptyPattern          = errors.New("input pattern was empty string")
+	errInvalidWildcard       = errors.New("wildcard '*' can only appear between two separators")
+	errInvalidDoubleWildcard = errors.New("wildcard '**' can only appear at end of pattern")
+	errRegexpCompile         = errors.New("unable to compile generated regex (internal bug)")
+)
+
+// Matcher represents a compiled pattern that can be matched against a string.
+type Matcher interface {
+	// Matches checks if the given input matches the compiled pattern.
+	Matches(string) bool
+}
+
+type regexpMatcher struct {
+	// Compiled regular expression for this matcher
+	pattern *regexp.Regexp
+}
+
+// Compile creates a new Matcher given a pattern, using '/' as the separator.
+func Compile(pattern string) (Matcher, error) {
+	return CompileWithSeparator(pattern, defaultSeparator)
+}
+
+// CompileList creates new Matchers given a list patterns, using '/' as the separator.
+func CompileList(patterns []string) ([]Matcher, error) {
+	ms := []Matcher{}
+	for _, pattern := range patterns {
+		m, err := Compile(pattern)
+		if err != nil {
+			return nil, err
+		}
+		ms = append(ms, m)
+	}
+	return ms, nil
+}
+
+// MustCompile creates a new Matcher given a pattern, using '/' as the separator,
+// and panics if the given pattern was invalid.
+func MustCompile(pattern string) Matcher {
+	m, err := CompileWithSeparator(pattern, defaultSeparator)
+	if err != nil {
+		panic(err)
+	}
+	return m
+}
+
+// CompileWithSeparator creates a new Matcher given a pattern and separator rune.
+func CompileWithSeparator(pattern string, separator rune) (Matcher, error) {
+	// Build regular expression from wildcard pattern
+	// - Wildcard '*' should match all chars except forward slash
+	// - Wildcard '**' should match all chars, including forward slash
+	// All other regex meta chars will need to be quoted
+
+	if pattern == "" {
+		return nil, errEmptyPattern
+	}
+
+	segments := strings.Split(pattern, string(separator))
+
+	var regex bytes.Buffer
+	regex.WriteString("^")
+
+loop:
+	for i, segment := range segments {
+		switch segment {
+		case "*":
+			// Segment with wildcard
+			regex.WriteString("[^")
+			regex.WriteRune(separator)
+			regex.WriteString("]+")
+		case "**":
+			// Segment with double wildcard
+			// May only appear at the end of a pattern
+			if i != len(segments)-1 {
+				return nil, errInvalidDoubleWildcard
+			}
+			regex.WriteRune('?')
+			regex.WriteString(".*$")
+			break loop
+		default:
+			// Segment to match literal string
+			if strings.Contains(segment, "*") {
+				return nil, errInvalidWildcard
+			}
+			regex.WriteString(regexp.QuoteMeta(segment))
+		}
+
+		// Separate this segment from next one
+		regex.WriteRune(separator)
+
+		if i == len(segments)-1 {
+			// Final slash should be optional
+			// We want "path" and "path/" to match
+			regex.WriteString("?$")
+		}
+	}
+
+	compiled, err := regexp.Compile(regex.String())
+	if err != nil {
+		return nil, errRegexpCompile
```

Comment:
Yeah I considered that but I think it's better if a bug in the library just leads to an error (which must be handled anyway) instead of a panic for the entire app.